File: code/dart/modular_api/test/tracing/tracing_pipeline_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:dartastic_opentelemetry/dartastic_opentelemetry.dart' as sdk;
import 'package:dartastic_opentelemetry/testing.dart';
import 'package:http/http.dart' as http;
import 'package:modular_api/modular_api.dart';
import 'package:test/test.dart';

/// Integration tests for tracing as the host actually wires it.
///
/// The middleware's own behaviour is covered in `tracing_middleware_test.dart`.
/// What is tested here is the wiring: that `serve()` installs it in the right place,
/// that absent options install nothing at all, and that operational routes stay out
/// of the trace store.
///
/// These run a real server on a real port, because the thing under test is the
/// pipeline `serve()` builds — asserting it by inspecting a `Pipeline` object would
/// be testing a reconstruction of the code rather than the code.
void main() {
  late TestHarness harness;
  late InMemorySpanExporter spans;

  setUpAll(() async {
    harness = await maybeInitializeOtelForTest(serviceName: 'modular_api-test');
    spans = harness.spans;
  });

  setUp(() => harness.clear());

  TracingOptions tracingOptions() =>
      TracingOptions(tracerProvider: sdk.OTel.tracerProvider());

  ModularApi apiWith({TracingOptions? tracing}) {
    final api = ModularApi(
      basePath: '/api',
      title: 'modular_api-test',
      tracing: tracing,
    );
    api.module('cuenta', (m) {
      m.usecase(
        'ping',
        PingUseCase.fromJson,
        inputExample: PingInput(value: 'ping'),
        outputExample: PingOutput(echo: 'ping'),
      );
      m.usecase(
        'boom',
        BoomUseCase.fromJson,
        inputExample: PingInput(value: 'ping'),
        outputExample: PingOutput(echo: 'ping'),
      );
    });
    return api;
  }

  Future<http.Response> post(int port, String path) => http.post(
        Uri.parse('http://localhost:$port$path'),
        headers: {'content-type': 'application/json'},
        body: jsonEncode({'value': 'ping'}),
      );

  group('with TracingOptions', () {
    late HttpServer server;

    tearDown(() async => server.close(force: true));

    test('a request through the real pipeline produces a server span', () async {
      server = await apiWith(tracing: tracingOptions()).serve(port: 0);

      await post(server.port, '/api/cuenta/ping');

      expect(spans.spanNames, contains('POST /api/cuenta/ping'));
    });

    test('the span carries the route and a 200 status', () async {
      server = await apiWith(tracing: tracingOptions()).serve(port: 0);

      await post(server.port, '/api/cuenta/ping');

      final span = spans.findSpanByName('POST /api/cuenta/ping')!;
      expect(span.attributes.getString('url.path'), equals('/api/cuenta/ping'));
      expect(span.attributes.getInt('http.response.status_code'), equals(200));
    });

    test('operational routes produce no span', () async {
      // Health, docs, openapi and metrics are noise in a trace store. The host
      // derives the exclusion list from operationalRoutePaths rather than hardcoding
      // it, so this also guards against that list drifting.
      server = await apiWith(tracing: tracingOptions()).serve(port: 0);

      await http.get(Uri.parse('http://localhost:${server.port}/api/health'));
      await http.get(Uri.parse('http://localhost:${server.port}/api/openapi.json'));

      expect(spans.spans, isEmpty);
    });

    test('a failing use case still produces an ended span', () async {
      server = await apiWith(tracing: tracingOptions()).serve(port: 0);

      await post(server.port, '/api/cuenta/boom');

      final span = spans.findSpanByName('POST /api/cuenta/boom');
      expect(span, isNotNull);
      expect(span!.endTime, isNotNull);
    });

    test('the span is a child of an incoming traceparent', () async {
      // End to end, through a real socket: the header arrives, propagation resolves
      // it, and the span attaches. This is the path Cloud Run exercises.
      server = await apiWith(tracing: tracingOptions()).serve(port: 0);
      const traceId = '4bf92f3577b34da6a3ce929d0e0e4736';

      await http.post(
        Uri.parse('http://localhost:${server.port}/api/cuenta/ping'),
        headers: {
          'content-type': 'application/json',
          'traceparent': '00-$traceId-00f067aa0ba902b7-01',
        },
        body: jsonEncode({'value': 'ping'}),
      );

      final span = spans.findSpanByName('POST /api/cuenta/ping')!;
      expect(span.spanContext.traceId.hexString, equals(traceId));
      expect(
        span.spanContext.parentSpanId?.hexString,
        equals('00f067aa0ba902b7'),
      );
    });
  });

  group('without TracingOptions (gate G3)', () {
    late HttpServer server;

    tearDown(() async => server.close(force: true));

    test('no span is produced at all', () async {
      // The structural half of G3: off means nothing installed, not "installed and
      // idle". Even a no-op tracer would have produced span objects here.
      server = await apiWith().serve(port: 0);

      final response = await post(server.port, '/api/cuenta/ping');

      expect(response.statusCode, equals(200));
      expect(spans.spans, isEmpty);
    });

    test('the API still serves normally', () async {
      // Invariant 3: a REST-only API is valid with or without optional plugins.
      server = await apiWith().serve(port: 0);

      final response = await post(server.port, '/api/cuenta/ping');

      expect(response.statusCode, equals(200));
      expect(jsonDecode(response.body), equals({'echo': 'ping'}));
      // The home-grown correlation header is untouched by tracing being absent.
      expect(response.headers['x-request-id'], isNotNull);
    });
  });
}

// ─── Fixtures ──────────────────────────────────────────────────────

class PingInput extends Input {
  PingInput({required this.value});

  factory PingInput.fromJson(Map<String, dynamic> json) =>
      PingInput(value: json['value'] as String? ?? '');

  final String value;

  @override
  Map<String, dynamic> toJson() => {'value': value};

  @override
  Map<String, dynamic> toSchema() => {
        'type': 'object',
        'properties': {
          'value': {'type': 'string'},
        },
        'required': ['value'],
      };
}

class PingOutput extends Output {
  PingOutput({required this.echo});

  final String echo;

  @override
  int get statusCode => 200;

  @override
  Map<String, dynamic> toJson() => {'echo': echo};

  @override
  Map<String, dynamic> toSchema() => {
        'type': 'object',
        'properties': {
          'echo': {'type': 'string'},
        },
        'required': ['echo'],
      };
}

class PingUseCase implements UseCase<PingInput, PingOutput> {
  PingUseCase({required this.input});

  static PingUseCase fromJson(Map<String, dynamic> json) =>
      PingUseCase(input: PingInput.fromJson(json));

  @override
  final PingInput input;

  @override
  ModularLogger? logger;

  @override
  String? validate() => null;

  @override
  Future<PingOutput> execute() async => PingOutput(echo: input.value);
}

class BoomUseCase implements UseCase<PingInput, PingOutput> {
  BoomUseCase({required this.input});

  static BoomUseCase fromJson(Map<String, dynamic> json) =>
      BoomUseCase(input: PingInput.fromJson(json));

  @override
  final PingInput input;

  @override
  ModularLogger? logger;

  @override
  String? validate() => null;

  @override
  Future<PingOutput> execute() async => throw StateError('boom');
}

File: code/ts/modular_api/test/tracing/tracingPipeline.test.ts
import { InMemorySpanExporter, SimpleSpanProcessor } from '@opentelemetry/sdk-trace-base';
import { NodeTracerProvider } from '@opentelemetry/sdk-trace-node';
import type { Server } from 'node:http';
import request from 'supertest';
import { afterEach, beforeAll, beforeEach, describe, expect, it } from 'vitest';

import { Input, ModularApi, Output, TracingOptions, UseCase } from '../../src';

/**
 * The full pipeline, through a real request.
 *
 * TypeScript mirror of `test/tracing/tracing_pipeline_test.dart`. The middleware tests exercise
 * `tracingMiddleware` directly; this exercises what `ModularApi.serve()` actually assembles, which is
 * the only thing that can catch the host wiring the middleware in the wrong place or not at all.
 *
 * **The G3 group is why this file was added at the merge gate.** Dart asserted the structural half of
 * G3 — that absent `tracing` means nothing is installed, not "installed and idle" — through a real
 * request. TypeScript and Python asserted G3 only at the middleware and satellite level, so the host's
 * conditional was never exercised in either. D1 says a stage advances all three; this one had not.
 */
class PingInput implements Input {
  toJson() {
    return {};
  }
  toSchema() {
    return { type: 'object', properties: {} };
  }
}

class PingOutput implements Output {
  get statusCode() {
    return 200;
  }
  toJson() {
    return { echo: 'ping' };
  }
  toSchema() {
    return { type: 'object', properties: { echo: { type: 'string' } } };
  }
}

class PingUseCase extends UseCase<PingInput, PingOutput> {
  readonly input: PingInput;

  constructor(input: PingInput) {
    super();
    this.input = input;
  }

  static fromJson(_json: Record<string, unknown>) {
    return new PingUseCase(new PingInput());
  }

  validate() {
    return null;
  }

  async execute() {
    return new PingOutput();
  }
}

const exporter = new InMemorySpanExporter();
let provider: NodeTracerProvider;

beforeAll(() => {
  provider = new NodeTracerProvider({ spanProcessors: [new SimpleSpanProcessor(exporter)] });
  provider.register();
});

beforeEach(() => exporter.reset());

let server: Server | undefined;

afterEach(async () => {
  if (server === undefined) return;
  const closing = server;
  server = undefined;
  await new Promise<void>((resolve, reject) =>
    closing.close((error) => (error ? reject(error) : resolve())),
  );
});

function apiWith(tracing?: TracingOptions): ModularApi {
  const api = new ModularApi({
    basePath: '/api',
    title: 'modular_api-test',
    ...(tracing === undefined ? {} : { tracing }),
  });
  api.module('cuenta', (m) => {
    m.usecase('ping', PingUseCase.fromJson, {
      inputClass: PingInput,
      outputClass: PingOutput,
    });
  });
  return api;
}

const spanNames = () => exporter.getFinishedSpans().map((span) => span.name);

describe('with TracingOptions', () => {
  it('a request through the real pipeline produces a server span', async () => {
    server = await apiWith(new TracingOptions({ tracerProvider: provider })).serve({ port: 0 });

    await request(server).post('/api/cuenta/ping').send({});

    expect(spanNames()).toContain('POST /api/cuenta/ping');
  });

  it('the span carries the route and a 200 status', async () => {
    server = await apiWith(new TracingOptions({ tracerProvider: provider })).serve({ port: 0 });

    await request(server).post('/api/cuenta/ping').send({});

    const span = exporter.getFinishedSpans().find((s) => s.name === 'POST /api/cuenta/ping');
    expect(span?.attributes['url.path']).toBe('/api/cuenta/ping');
    expect(span?.attributes['http.response.status_code']).toBe(200);
  });

  it('operational routes produce no span', async () => {
    // Health, docs, openapi and metrics are noise in a trace store. The host derives the exclusion
    // list from the operational paths rather than hardcoding it, so this also guards that list
    // against drift.
    server = await apiWith(new TracingOptions({ tracerProvider: provider })).serve({ port: 0 });

    await request(server).get('/api/health');

    expect(exporter.getFinishedSpans()).toHaveLength(0);
  });
});

describe('without TracingOptions (gate G3)', () => {
  it('no span is produced at all', async () => {
    // The structural half of G3: off means nothing installed, not "installed and idle". Even a no-op
    // tracer would have produced span objects here.
    server = await apiWith().serve({ port: 0 });

    const response = await request(server).post('/api/cuenta/ping').send({});

    expect(response.status).toBe(200);
    expect(exporter.getFinishedSpans()).toHaveLength(0);
  });

  it('the API still serves normally', async () => {
    // Invariant 3: a REST-only API is valid with or without optional plugins.
    server = await apiWith().serve({ port: 0 });

    const response = await request(server).post('/api/cuenta/ping').send({});

    expect(response.status).toBe(200);
    expect(response.body).toEqual({ echo: 'ping' });
    // The home-grown correlation header is untouched by tracing being absent.
    expect(response.headers['x-request-id']).toBeDefined();
  });
});
